smt.go

```diff
@@ -64,75 +64,31 @@ func (smt *SparseMerkleTree) depth() int {
 	return smt.th.pathSize() * 8
 }
 
-// Get gets a key from the tree.
+// Get gets the value of a key from the tree.
 func (smt *SparseMerkleTree) Get(key []byte) ([]byte, error) {
-	value, err := smt.GetForRoot(key, smt.Root())
-	return value, err
-}
+	// Get tree's root
+	root := smt.Root()
 
-// GetForRoot gets a key from the tree at a specific root.
-func (smt *SparseMerkleTree) GetForRoot(key []byte, root []byte) ([]byte, error) {
 	if bytes.Equal(root, smt.th.placeholder()) {
 		// The tree is empty, return the default value.
 		return defaultValue, nil
 	}
 
 	path := smt.th.path(key)
-	currentHash := root
-	for i := 0; i < smt.depth(); i++ {
-		currentData, err := smt.nodes.Get(currentHash)
-		if err != nil {
-			return nil, err
-		} else if smt.th.isLeaf(currentData) {
-			// We've reached the end. Is this the actual leaf?
-			p, _ := smt.th.parseLeaf(currentData)
-			if !bytes.Equal(path, p) {
-				// Nope. Therefore the key is actually empty.
-				return defaultValue, nil
-			}
-			// Otherwise, yes. Return the value.
-			value, err := smt.values.Get(path)
-			if err != nil {
-				return nil, err
-			}
-			return value, nil
-		}
-
-		leftNode, rightNode := smt.th.parseNode(currentData)
-		if getBitAtFromMSB(path, i) == right {
-			currentHash = rightNode
-		} else {
-			currentHash = leftNode
-		}
-
-		if bytes.Equal(currentHash, smt.th.placeholder()) {
-			// We've hit a placeholder value; this is the end.
-			return defaultValue, nil
-		}
-	}
-
-	// The following lines of code should only be reached if the path is 256
-	// nodes high, which should be very unlikely if the underlying hash function
-	// is collision-resistant.
 	value, err := smt.values.Get(path)
 	if err != nil {
-		return nil, err
+		return defaultValue, nil
```

Comment:
One design question here is: `Get` and `Has` no longer have an error condition; they always succeed, since `Get` returns `defaultValue` is the key has default value. Should these functions return something like a `DefaultValueError` in such a case, or should they return no error at all?

Reply:
The check in the original version only errors if the underlying mapstore is somehow corrupted, and the code wasn't able to get a key that is expected to have a value.

defaultValue should always be returned if no value was found, otherwise.

Reply:
Hmmm. I think I should keep the error for now, but in a future PR we should add error handling to differentiate between a key-has-default-value "error" and a database error (since the underlying db, e.g. Badger or Rocks, might have different errors). How does that sound?

Reply:
What error? There's no error being returned in this code.

Reply:
I think the code needs to check what the error type is, and only return defaultValue is the value store returns an non-existent key error or similar.

Reply:
> What error? There's no error being returned in this code.

By "keep the error" I mean "keep the `error` return value," which is always `nil` for now. As opposed to removing the `error` and changing all the uses of `Get`/`Has` to not have two return values.

> I think the code needs to check what the error type is, and only return defaultValue is the value store returns an non-existent key error or similar.

Right, that's what

> I think I should keep the error for now, but in a future PR we should add error handling to differentiate between a key-has-default-value "error" and a database error (since the underlying db, e.g. Badger or Rocks, might have different errors). 

is trying to say!

Note that the current memory-backed mapstore implementation only has a single error when a key isn't found, so the PR as written handles that one and only case. I think handling of an additional error type (e.g. actual database failure) can be done in a separate PR.

Reply:
I think the code should at least check if it's an InvalidKeyError, and bubble the error otherwise.

Reply:
Actually that was easier than I thought. Fixed in e02b89f.